### ISSUE TITLE
CID Specification does not define "base58btc"

### ISSUE THREAD

**Issue by JDLH** [multiformats/cid#45]
The [CID Specification](https://github.com/multiformats/cid/blob/d394fdd49c568f055fb221160f7bbbd848480d47/README.md) refers to an encoding/decoding method "base58btc" but does not define what this method is. 

I suspect the intended meaning is "Base58 encoding as used by Bitcoin". It might also mean "Base58Check encoding as defined by Bitcoin.  Base58Check appears to have properties like: 1. well-defined way of handling leading zero bytes in the binary payload, and thus ability to represent any length of binary payload; 2. a notion of a one-byte "prefix" to which the binary payload is concatenated; 3. a 4-byte checksum which is concatenated to the prefix and payload. It is not clear to me whether or not the CID Specification's term "base58btc" is supposed to include these properties.

I am new to Multiformats and IPFS and Filecoin, so I have a very naive reading of the CID Specification. I did a little web searching, and could not find something which purported to be an authoritative spec for something named "base58btc". There are many pages which describe Base58 representation, and/or Bitcoin's Base58Check encoding, but none of these seemed to claim to be an authoritative specification. I may well have overlooked something.

[**The Base58 Encoding Scheme**](https://tools.ietf.org/html/draft-msporny-base58-02) is an IETF internet draft by M Sporny. It describes the Base58 alphabet used by Bitcoin, and property 1. above. It does not include properties 2. or 3. This draft is labelled as a "draft", and as expiring in May 2021, so it may not be robust enough to refer to in the CID Specification.

[**Base58check encoding**](https://www.reference.cash/protocol/blockchain/encoding/base58check/) from the Bitcoin Cash Protocol website appears to have a pretty clear specification for the Base58Check encoding as defined by Bitcoin. I don't know how authoritative it is.

I saw a piece of source code which implemented Base58Check encoding, which quote extensively from Bitcoin source code defining Base58Check encoding. The Bitcoin content was attributed to Satoshi Nakamoto, and was very clear. I don't know where to find an authoritative copy of the Bitcoin source code module containing this quote. Maybe that source code is a good spec to refer to from the CID Specification.

In any case, as a naive software developer attempting to understand the CID Specification, it would be helpful to have a clear reference to what is specified by "base58btc".

**Comment by vmx** [comment]
> I suspect the intended meaning is "Base58 encoding as used by Bitcoin".

Yes, that is the intended meaning. And the "[**The Base58 Encoding Scheme**](https://tools.ietf.org/html/draft-msporny-base58-02) is an IETF internet draft by M Sporny" would be the one to implement.

I know, it doesn't make up for a proper specification, but the Multibase repository contains some [test vectors](https://github.com/multiformats/multibase/tree/ae81809e22f9c8d8396e96b83c8bdae687a00c70/tests), in case you want to check an implementation against base58btc encoding.

I've also [opened an issue on the Multibase spec](https://github.com/multiformats/multibase/issues/76) as I think this would be the right place to add the information about which specification a Multibase is based upon.

**Comment by rvagg** [comment]
^ to reinforce @vmx' comment, the reason it's probably not clear is that we're all operating on a collection of assumed knowledge that shouldn't be assumed for the purpose of a spec! The string `base58btc` specifically comes from the multibase table, line 19: https://github.com/multiformats/multibase/blob/master/multibase.csv#L19
